Step2 리뷰 반영 & Step3

michaelkimm opened the pull request:
# 반영 사항
- [x] [이 구조라면, 단계가 넘어갈수록 (검증 요구사항이 많아질수록) If문이 늘어날 것 같아요.](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267765134)
    -> 구간 추가 핸들러 이용하여 '검증'과 '구간 추가' 로직 캡슐화

- [x] [Exception 메세지를 상수화해보면 어떨까요? ](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267763020)
    -> MessageSource 이용, 도메인은 예외 내부에 메세지 코드만 넣어서 예외 던짐

- [x] [Http응답을 검증하는 거라면, restAssured에서 바로 할 수도 있어요.](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267744505)

- [x] ["구간을 등록한다, 구간 등록 실패"가 prefix로 붙네요. @nested로 그룹핑 해주셔도 좋을 것 같아요.](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267735794)

- [ ] [다음 단계에서 Mockito를 이용한 지하철 구간 테스트 작성 부탁드려요!](https://github.com/next-step/atdd-subway-path/pull/610#issuecomment-1641707652)
    -> 미적용 이유: 지하철 구간 테스트 작성이 잘이해가 안됐습니다! Section을 테스트하는 것인가요? 아니면 Sections 혹은 Line인가요? '구간'이라면 Section인데 Section은 테스트할 것이 없어서 여쭤봅니다!

- [ ] [equals를 Override 해보는 건 어떨까요? ](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267749308)
    -> 미적용 이유: 댓글 참조바랍니다!

# 참고사항
1. 커밋 [feat: 구간 추가 핸들러 단위 테스트 작성](https://github.com/next-step/atdd-subway-path/pull/623/commits/0e350501094157bc0af12dc395831b5c66849371) 부터가 이번 PR에 적용된 작업 건입니다! 이전 커밋은 이미 머지한 건인데 나타나네요😅
2. 스탭3 브랜치인데 스탭2 리뷰 적용까지만 먼저하고 PR보내요! step3에서 할 것이 많아 보여서 step2 리뷰 부터 요청드립니다! 리뷰 완료 후 머지 안해주시면 스텝3도 이어서 현 PR에 커밋할께요🙂

michaelkimm committed:
#600 pr 리뷰 반영

1. sections의 필드를 직접 가져와서 비즈니스 로직을 검증하지 않고, sections의 메소드를 호출하는 방식으로 변경

michaelkimm committed:
test: 지하철 구간 추가 기능 개선 happy case 인수 테스트 인수 조건 정의

michaelkimm committed:
test: 지하철 구간 추가 기능 개선 happy case 인수 테스트 작성

michaelkimm committed:
test: 구간 등록 도메인 단위 테스트

michaelkimm committed:
feat: 단위 테스트를 통과하는 도메인 코드 작성

michaelkimm committed:
test: 노선 조회시 응답되는 역 목록 수정 도메인 단위 테스트 작성

michaelkimm committed:
feat: 인수 테스트 통과하는 도메인 로직 작성

구현 사항
1. 기존 구간 A-C에 신규 구간 A-B를 추가하는 경우 A역을 기준으로 추가
2. 새로운 역을 상행 종점역으로 등록할 경우의 구간 추가 성공 케이스
3. 노선 조회시 응답되는 역 목록 상행종착역-하행종착역 순으로 반환

michaelkimm committed:
feat: 새 구간 하행역 == 기존 구간 하행 종착역 경우 구간 등록 happy case 테스트 및 도메인 로직 구현

michaelkimm committed:
test: 변경된 구간 등록 예외 케이스 인수 테스트 추가

테스트 조건
1. 역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 등록을 할 수 없음
2. 상행역과 하행역이 이미 노선에 모두 등록되어 있다면 추가할 수 없음
3. 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음

michaelkimm committed:
test: 구간 등록 A-C에 B-C 추가 단위 테스트 기능 구현 후 작성

michaelkimm committed:
test: 변경된 구간 등록 예외 케이스 도메인 단위 테스트 추가

michaelkimm committed:
feat: 변경된 구간 등록 예외 케이스 도메인 단위 테스트 통과 로직 구현

michaelkimm committed:
feat: 예외 세분화

최상위 예외: Exception
차상위 예외: IllegalArgumentException, BusinessException
하위 예외: ValidationException

michaelkimm committed:
feat: 테스트 내 magic number 삭제

michaelkimm committed:
feat: 구간 추가 예외 케이스 추가

노선 A역-B역-C역 존재 시 B역-A역 추가 케이스 예외 처리

michaelkimm committed:
feat: 구간 추가 핸들러 단위 테스트 작성

michaelkimm committed:
feat: 구간 추가 핸들러 단위 테스트 완성

michaelkimm committed:
feat: 구간 추가 핸들러 단위 테스트 통과 로직 작성

michaelkimm committed:
feat: 책임 연쇄 패턴 적용하여 구간 추가 시 공통 검증 핸들러 캡슐화

michaelkimm committed:
ref: 필요 없는 주석 및 코드 삭제

michaelkimm committed:
ref: #610반영, RestAssured chaining으로 http 상태코드 곧바로 검증

michaelkimm committed:
ref: #610 반영, @Nest를 이용하여 같은 맥락의 테스트를 묶어줌.

michaelkimm committed:
test: 메세지소스 학습 테스트 작성

michaelkimm committed:
feat: 메세지 소스를 이용한 예외 메세지 상수화

michaelkimm committed:
feat: 구간추가핸들러매핑을 빈으로 변경

michaelkimm commented on src/main/java/nextstep/subway/line/entity/handler/SectionAdditionHandlerMapping.java:
```diff
@@ -3,19 +3,21 @@
 import java.util.List;
 
+@Component
 public class SectionAdditionHandlerMapping {
```
**고민한 point**
 `'어떤 방식으로 구간을 추가하는가' -> '검증' -> '구간 추가' `세가지 로직은 캡슐화 돼야한다고 판단했어요. [기존 코드](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267765134)에선 `Validator` static inner class가 존재해서 검증만 캡슐화가 되있기에 `'어떤 방식으로 구간을 추가하는가' -> '검증'`과 `'어떤 방식으로 구간을 추가하는가' -> '구간 추가'`가 따로 존재했습니다. 핸들러는 세가지를 하나로 묶어주는 역할을 합니다🙂

michaelkimm commented on src/main/java/nextstep/subway/line/entity/Sections.java:
```diff
@@ -55,8 +55,8 @@ public void forceSectionAddition(Section section) {
-    public void addSection(Section section) {
-        SectionAdditionHandler handler = SectionAdditionHandlerMapping.getHandler(this, section);
+    public void addSection(SectionAdditionHandlerMapping handlerMapping, Section section) {
+        SectionAdditionHandler handler = handlerMapping.getHandler(this, section);
```
구간 추가 핸들러 매핑을 static으로 사용하지 않고 빈으로 등록하는 방식으로 변경했습니다! 빈으로 매핑을 사용하면 도메인을 사용하는 사람이 `SectionAdditionHandlerMapping`를 사용해서 구간 추가를 해야하는 것을 명확히 알 수 있다고 판단했기 때문이에요(필요 없는 결합도가 생기지 않는가?🤔라는 의문이 생길 수 있지만, 도메인 객체와 구간추가핸들러매핑은 같은 도메인의 비즈니스 로직에 속하기에 결합도가 높아야한다고 생각했어요). 그에 비해 static을 사용하면 구현체를 들어가야 알 수 있기 때문에 좋지 않을거라 판단했어요🙂

michaelkimm committed:
Merge branch 'michaelkimm' into step3

michaelkimm committed:
err: 테스트 미통과 되는 에러 해결

michaelkimm committed:
test: 경로 조회 인수 테스트 작성

michaelkimm committed:
test: 최단 경로 조회 mock 이용한 단위 테스트 작성

hiblue02 commented on src/main/resources/messages_ko_KR.properties:
```diff
@@ -0,0 +1,13 @@
+station.down.not.found=\uD558\uD589\uC5ED\uC774 \uC874\uC7AC\uD558\uC9C0 \uC54A\uC2B5\uB2C8\uB2E4.
+station.top.not.found=\uC0C1\uD589 \uC885\uCC29\uC5ED\uC744 \uCC3E\uC744 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.
+station.last.not.found=\uD558\uD589 \uC885\uCC29\uC5ED\uC744 \uCC3E\uC744 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.
+station.name.already.exists=\uC774\uBBF8 \uC874\uC7AC\uD558\uB294 \uC5ED \uC774\uB984\uC785\uB2C8\uB2E4.
```
messages.properties를 써주셨네요! :+1:
다만 한글이 깨졌어요 😭

michaelkimm replied:
@hiblue02 인텔리제이 설정을 변경해주니 해결되네요! 감사합니다!

hiblue02 commented on src/main/java/nextstep/subway/line/entity/handler/SectionAdditionHandlerMapping.java:
```diff
@@ -0,0 +1,28 @@
+            new AddSectionAtFirstHandler(new AnyStationPreExistCheckHandler(null)),
+            new AddSectionAtMiddleRightHandler(new AnyStationPreExistCheckHandler(null)),
+            new AddSectionAtMiddleLeftHandler(new AnyStationPreExistCheckHandler(null))
+    );
```
전략 패턴을 사용해 개선해주셨네요! :100:

michaelkimm replied:
@hiblue02 혜영님 덕분입니다!:)

hiblue02 commented on src/test/java/nextstep/subway/unit/LineServiceMockTest.java:
```diff
@@ -33,7 +37,7 @@ void addSection() {
         Line line = mock(Line.class);
-        LineService lineService = new LineService(lineRepository, stationRepository);
+        LineService lineService = new LineService(lineRepository, stationRepository, sectionAdditionHandlerMapping);
         when(stationRepository.findById(강남역.getId())).thenReturn(Optional.of(강남역));
```
Mockito 테스트 구현해주셧네요! sections 서비스에 대한 테스트를 요청드린 거였어요! 💯

michaelkimm replied:
@hiblue02 구현 했던걸로 이해하면 되는거죠!? 감사합니다:) 😄

hiblue02 reviewed (changes requested):
리뷰를 꼼꼼히 반영해주셨네요! 💯 
요청해주신 데로 머지하지 않겠습니다. 요 pr에 이어서 3단계도 진행 부탁드려요! 

그리고 equals에 대한 답변이 안보이네요. 😅  다음 리뷰요청에 추가해주세요!

michaelkimm committed:
feat: 최단 경로 조회 테스트 통과하는 서비스 로직 작성

michaelkimm committed:
fix: 메세지 한글 깨짐 해결

michaelkimm committed:
fix: 필요 없는 메세지 삭제

michaelkimm committed:
ref: equalsId -> equals로 변경

michaelkimm committed:
test: 최단 경로 알고리즘 단위 테스트 추가

michaelkimm commented on src/main/java/nextstep/subway/station/entity/Station.java:
```diff
@@ -24,7 +24,16 @@ public Station(String name) {
+        if (this == o) return true;
+        if (!(o instanceof Station)) return false;
+        Station station = (Station) o;
+        return Objects.equals(id, station.getId()) && Objects.equals(name, station.getName());
```
@hiblue02 원래 객체간 비교할 때 id만 비교해도 되지 않나 생각했습니다! 모든 값을 다 비교하면 객체 내 새로운 필드를 추가할 때 마다 equals를 오버라이드 해줘야하는 번거로움이 있기 때문이죠! 하지만 혜영님의 [피드백](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267749308) 이후로 생각해보니, 아이디 값만 같고 값이 다른 객체는 다르다고 하는 것이 맞다고 생각했어요! 이에 equals로 객체간 비교 방식을 바꿨습니다!

 equals를 도입하는 과정에서 lazyloading으로 인한 프록시 객체의 동작 방식을 명확하게 이해하지 못해서 테스트할 때 많이 헤맸지만, [JPA Hibernate 프록시 제대로 알고 쓰기](https://tecoble.techcourse.co.kr/post/2022-10-17-jpa-hibernate-proxy/)를 읽은 후 equals 내에서 비교 대상 객체의 필드 값을 getter로 접근하도록 바꿔서 문제를 해결했습니다:)

hiblue02 replied:
우와 공부 많이 하셨군요! 💯

michaelkimm committed:
test: pathServiceMock 단위 테스트 재작성

michaelkimm reviewed (commented):
# 반영 사항

- [x] [한글 깨짐 현상](https://github.com/next-step/atdd-subway-path/pull/623#discussion_r1270639045)
- [x] [equalsId -> equals](https://github.com/next-step/atdd-subway-path/pull/610#discussion_r1267749308)

# 질문
- 수업에서 의도한 가짜 협력 객체 사용 방식이 어떤 것인지 모르겠습니다😂 [넥스트 스텝 3단계 요구 사항 명세서](https://edu.nextstep.camp/s/c30VOHNP/ls/yamKKPa1https://edu.nextstep.camp/s/c30VOHNP/ls/yamKKPa1)에서는 Outside In 방식을 설명하면서 `외부 라이브러리를 활용한 로직을 검증할 때는 가급적 실제 객체를 활용`라고 언급했습니다. 그렇다면 서비스 레이어 내에 외부 라이브러리를 활용한 로직이 있을 땐 이와 관련된 객체는 실제 객체를 사용하는 것을 말하는건가요? 이 말이 맞다면 테스트는 더 세부적으로 나눠야하지 않나요?🤔 예를 들어, `서비스가 외부 라이브러리를 호출하는 것을 확인하는 테스트`와 `외부라이브러리 그 자체가 잘 동작하는지 테스트` 두가지로요. 처음에는 [pathService 단위 테스트에서 shortestPathFinder 실제 객체를 이용](https://github.com/next-step/atdd-subway-path/pull/623/commits/36b5a0a1e5b0c3461ea564d9b0d760560a1c8225)해서 테스트를 진행했다가 테스트를 분리해야겠다고 생각해서 추후에 나누었는데([pathservice 단위 테스트](https://github.com/next-step/atdd-subway-path/pull/623/commits/00ab9245b0b9a56c6f2de260fd0d433d7b15ba27), [shortestPath 단위 테스트](https://github.com/next-step/atdd-subway-path/pull/623/commits/b27889ddf48ad2b34c044774bccddbbe6333cffe)), 혜영님의 생각이 궁금하네요🙂

> 넥스트 스텝 페이지 3단계 마지막에 적힌 글
 두 방향성을 모두 사용해보시고 테스트가 협력 객체의 세부 구현에 의존하는 경우(가짜 협력 객체 사용)와 테스트 대상이 협력 객체와 독립적이지 못하고 변경에 영향을 받는 경우(실제 협력 객체 사용)를 모두 경험해보세요 :)

hiblue02 commented on src/main/java/nextstep/subway/line/entity/Sections.java:
```diff
@@ -101,13 +105,17 @@ else if (checkDownStationsContains(station)) {
         }
```
구간 삭제도 addSectionHandler처럼 전략패턴을 적용해볼 수도 있지 않을까요? :thinking:

michaelkimm replied:
@hiblue02 구간 삭제는 현재로선 다양한 삭제 케이스가 존재하지 않아서 if 문으로 남겨뒀습니다! 그래도 추후 유지 보수에는 전략 패턴을 적용하는 것이 좋을 듯 하네요 ㅎㅎ 적용해보겠습니다!

hiblue02 commented on src/main/java/nextstep/subway/line/service/PathService.java:
```diff
@@ -0,0 +1,45 @@
+        List<Line> lineList = lineRepository.findAll();
+
+
+        List<Station> searchedPath = shortestPathFinder.searchPath(lineList, sourceStation, targetStation);
+        BigInteger weight = shortestPathFinder.getShortestWeight(lineList, sourceStation, targetStation);
+        return new ShortestPathResponse(searchedPath.stream().map(StationResponse::from).collect(Collectors.toList()), weight);
+    }
```
최단경로와 거리를 구하기 위해 lineList와 sourceStation, targetStation을 각각 세팅하고 있는데요. 이러면 휴먼에러가 발생할 수도 있을 것 같아요.
shortestPathFinder의 최단경로와 거리를 구하는 코드도 거의 비슷한데, 좀 더 좋은 방법이 있을 것 같아요. :smile:
`경로`를 담당하는 도메인 객체가 있어도 좋을 것 같네요.

michaelkimm replied:
@hiblue02 아주 동의합니다! 안그래도 `ShortestPathFinder` 내부에 코드 반복이 존재해서 꺼림찍했네요 ㅎㅎ 다만 도메인 객체가 무엇인지는 생각해볼 주제네요. 저는 이때까지 비즈니스 로직의 핵심 부분을 담당하는 `@Entity`를 도메인 객체라고 판단했어요. 근데 혜영님의 말을 듣고 [Domain Model - Martin Fowler](https://martinfowler.com/eaaCatalog/domainModel.html)를 읽고 보니, 도메인 객체는 비즈니스 로직의 핵심 개념을 코드로 옮긴 개념이라고 생각이 바뀌었습니다! **혜영님도 이를 의도하신건가요?** 예를들어 제가 만든 `AddSectionHandler`는 도메인 객체가 되는거죠!

hiblue02 replied:
AddSectionHandler는 도메인 서비스에 가까운 것 같아요. 
https://incheol-jung.gitbook.io/docs/study/ddd-start/7

리뷰에 객체라고 적은 것은 도메인 객체가 아닌 자바의 Object 좀더 좁게는 클래스를 의미하는 거였어요.

hiblue02 commented on src/main/java/nextstep/subway/line/service/PathService.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final LineRepository lineRepository;
+    private final StationRepository stationRepository;
+    private final ShortestPathFinder shortestPathFinder;
```
PathService와 jgraph 라이브러리를 호출하는 부분을 분리해주셨네요! :+1:

hiblue02 commented:
서비스 레이어 내에 외부 라이브러리를 활용한 로직이 있을 땐 이와 관련된 객체는 실제 객체를 사용하는 것을 말하는건가요?
> 네! 맞아요. 😄 그래서 말씀해주신 바와 같이 외부 라이브러리 자체를 테스트하는 것과 서비스가 외부 라이브러리를 호출했는지 검증하는 테스트를 분리해야 된다고 생각합니다. 

1. `외부라이브러리 그 자체가 잘 동작하는지 테스트 `는 Outside In 흐름과 분리해야 된다고 생각해요. 그래서 이때는 외부 라이브러리의 동작을 검증하는 테스트 코드를 따로 작성합니다. 샘플코드의 [JgraphtTest](https://github.com/next-step/atdd-subway-path/blob/master/src/test/java/nextstep/study/JgraphtTest.java) 처럼요! (이걸로도 충분하다고 생각해요.)
2. 1번에서 라이브러리가 잘동작하는지 확인했으니, `서비스가 외부 라이브러리를 호출하는 것을 확인하는 테스트` 를 수행할 때는, 외부 라이브러리를 호출하는 실제 객체를 넣어서 테스트하면 된다고 생각해요. 원하는 형태로 결과를 만들어줬는지 검증하기 위해서요. 
 > 작성해주신 PathSerivceMockTest는 Mock을 써서 메소드 호출 결과를 모두 정해주고 있는데요. 저는 요게 의미 있는(?) 테스트가 아니라는 생각이 들어요. 그래서 서비스에서 외부 라이브러리 호출하는 부분에 대한 테스트는 실제 객체를 넣어서 검증해야 한다고 생각해요. 😄

hiblue02 reviewed (changes requested):
3단계 미션도 빠르게 해주셨네요! 👍 
코멘트 남겼는데 확인해주시고, 다시 리뷰요청 주세요!

hiblue02 commented on src/test/java/nextstep/subway/acceptance/path/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,76 @@
+import static org.assertj.core.api.Assertions.*;
+
+@DisplayName("지하철 경로 탐색")
+public class PathAcceptanceTest extends AcceptanceTest {
```
실패나, "역 개수는 많지만, 더 가까운 거리의 경로를 선택한다" 등 다양한 케이스의 테스트가 추가되면 좋을 것 같아요. :smile:

michaelkimm commented:
@hiblue02 [리뷰](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1646727944) 감사합니다! 

> [[Ref]](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1646727944) 작성해주신 PathSerivceMockTest는 Mock을 써서 메소드 호출 결과를 모두 정해주고 있는데요. 저는 요게 의미 있는(?) 테스트가 아니라는 생각이 들어요. 

맞네요 돌이켜보니 테스트의 목정성이 뚜렷하지 않네요 ㅎㅎ 

> [[Ref]](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1646727944) 1번에서 라이브러리가 잘동작하는지 확인했으니, 서비스가 외부 라이브러리를 호출하는 것을 확인하는 테스트 를 수행할 때는, 외부 라이브러리를 호출하는 실제 객체를 넣어서 테스트하면 된다고 생각해요. 원하는 형태로 결과를 만들어줬는지 검증하기 위해서요.

`원하는 형태로 결과를 만들어 줬는지 검증하는 것`이 PathService 단위 테스트의 주 목적이니, 함수 호출 여부를 검증하는 것이 아닌 Input값에 대해 예상한 Output 값이 나오는지 검증해야 하고 이에 실제 객체를 써야한다는 뜻인가요?!

michaelkimm committed:
test: PathService 단위 테스트 시, 실제 객체 사용

michaelkimm committed:
feat: PathFinder 도메인 객체 단위 테스트 및 통과 로직 작성

michaelkimm committed:
test: 지하철 경로 조회 실패 인수 테스트 작성

michaelkimm committed:
test: 최단 경로 조회 실패 Mock 단위 테스트 작성

michaelkimm committed:
feat: 최단 경로 조회 단위 테스트 통과 로직 작성

michaelkimm committed:
test: delete 핸들러 매핑 단위 테스트 뼈대 코드 작성

michaelkimm committed:
test: delete 핸들러 매핑 단위 테스트 코드 작성

michaelkimm committed:
feat: 구간 삭제 핸들러 매핑 테스트 통과 코드 작성

michaelkimm committed:
test: 구간 제거 핸들러 단위 테스트 추가

michaelkimm committed:
feat: sectionDeletionFilter와 SectionDeletionHandlerMapping을 호출하는 SectionDeletionOperator추가

1. 구간 삭제 시 공통 검증 부분이 추가됐어야함. 이에 필터 기능이 필요.
2. 핸들러는 각각의 구각 삭제 경우 따라 다르게 동작하는 객체임. 따라서 필터와 구분되야함.

michaelkimm commented:
# 반영 사항

- [x] [경로를 담당하는 도메인 객체가 있어도 좋을 것 같네요.](https://github.com/next-step/atdd-subway-path/pull/623#discussion_r1271369763)
- [x] [구간 삭제도 addSectionHandler처럼 전략패턴을 적용해볼 수도 있지 않을까요?](https://github.com/next-step/atdd-subway-path/pull/623#discussion_r1271369360)
- [x] [서비스에서 외부 라이브러리 호출하는 부분에 대한 테스트는 실제 객체를 넣어서 검증해야 한다고 생각해요. 😄](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1646727944)

_ps. 마지막에 생각한 [(출처)](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1646744866)`원하는 형태로 결과를 만들어 줬는지 검증하는 것이 PathService 단위 테스트의 주 목적이니, 함수 호출 여부를 검증하는 것이 아닌 Input값에 대해 예상한 Output 값이 나오는지 검증해야 하므로 실제 객체를 써야한다는 뜻인가요?!`를 바탕하여 먼저 코드를 작성했습니다!_

# 질문
- 구간 삭제 시 핸들러 매핑이 아닌 `SectionDeletionOperator`를 도메인 객체에서 호출하는 방식으로 변경했습니다. `Opertator`는 필터와 핸들러 매핑을 작동시킵니다. 필터는 공통 검증 부분을 처리하고 핸들러 매핑은 각 구간 삭제 케이스를 처리합니다. 혜영님의 리뷰 [구간 삭제도 addSectionHandler처럼 전략패턴을 적용해볼 수도 있지 않을까요?](https://github.com/next-step/atdd-subway-path/pull/623#discussion_r1271369360)를 보고 설계를 하다 보니 공통 검증 부분은 핸들러 매핑에서 따로 빼야겠더라구요. 근데 이 방식으로 하니 코드가 너무 흩어지는 느낌이 들었습니다. 전략패턴의 단점일까요. 단 확장성은 확실히 좋은 듯 합니다. **혜영님은 어떻게 생각하시나요?** 이렇게 까지 책임을 분할한 적은 처음이라 이래도 되나? 싶네요 ㅎㅎ (혜영님의 생각을 들어보고 전체 코드 구조를 바꿔보자고 생각하여 구간 추가 부분은 필터를 적용 안한 상태입니다)

_ps. 혜영님과 코드 리뷰를 하니 지나칠 만한 것이나 스스로 당연하다고 생각한 것도 한번 더 고민할 수 있어서 재밌습니다. 감사합니다😄_

michaelkimm committed:
test: SectionDeletionOpertor 단위 테스트 추가

hiblue02 commented:
> PathService 단위 테스트의 주 목적이니, 함수 호출 여부를 검증하는 것이 아닌 Input값에 대해 예상한 Output 값이 나오는지 검증해야 하고 이에 실제 객체를 써야한다는 뜻인가요?!

네! 요런 의도로 말씀드렸어요. 😄

hiblue02 commented:
> * 구간 삭제 시 핸들러 매핑이 아닌 `SectionDeletionOperator`를 도메인 객체에서 호출하는 방식으로 변경했습니다. `Opertator`는 필터와 핸들러 매핑을 작동시킵니다. 혜영님의 리뷰 https://github.com/next-step/atdd-subway-path/pull/623#discussion_r1271369360를 보고 설계를 하다 보니 공통 검증 부분은 핸들러 매핑에서 따로 빼야겠더라구요. 근데 이 방식으로 하니 코드가 너무 흩어지는 느낌이 들었습니다.

전 상행 종점역을 삭제할 때, 하행 종점역을 삭제할 때, 사이 역(구간)을 삭제할 때와 같이 비즈니스 로직이 명확하게 다른 부분에 대해서만 전략 패턴을 적용하면 된다고 생각해요. 말씀하신 것처럼 공통 검증 부분까지 클래스로 빼면, 오히려 로직이 너무 흩어지고 관리하기 어려울 것 같네요. (전략 패턴이 달라지는 부분을 추상화해서 갈아끼는(?) 디자인 패턴이기도 하니까요. 😄)

michaelkimm committed:
doc: git ignorea에 bin파일 추가

michaelkimm committed:
doc: 빈 파일 삭제

michaelkimm committed:
ref: 비즈니스 로직이 명확히 다른 부분만 전략 패턴 적용

1. 필터 삭제
2. 공통 검증 로직은 sections에서 핸들러 전에 처리

michaelkimm commented:
# 반영 사항

- [x] [공통 검증 부분까지 클래스로 빼면, 오히려 로직이 너무 흩어지고 관리하기 어려울 것 같네요. ](https://github.com/next-step/atdd-subway-path/pull/623#issuecomment-1647872468)
    -> 공통 검증 부분은 sections에서 처리, 추가&삭제 경우에 따라 비즈니스 로직이 달라지는 부분만 전략 패턴 적용

hiblue02 reviewed (approved):
민석님! 인수테스트와 TDD 미션 수행하느라 수고하셨습니다. 😄 
꼼꼼히 리뷰 반영해주시고, 질문도 해주셔서 리뷰 과정이 즐거웠어요!
다음 단계도 화이팅입니다. 🔥
